### PR TITLE
[#25] Implement /api/rest/company endpoint

### LINKED ISSUE #25: Create /api/rest/company/ endpoint

### PR DESCRIPTION
**Pull request by marsofandrew**
close #25

### CODE REVIEW

File: api/src/main/java/org/confeti/controllers/ControllersUtils.java [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 public final class ControllersUtils {
     public static final String REST_API_PATH = "/api/rest";
+    public static final String CONFERENCE_NAME_URI_PARAMETER = "conference_name";
+    public static final String YEAR_URI_PARAMETER = "year";
+    public static final String COMPANY_PARAMETER = "company_name";
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/ControllersUtils.java]: You named the rest of the contants as `SMTH_URI_PARAMETER`. Maybe you should name this constant `COMPANY_NAME_URI_PARAMETER`.

File: api/src/main/java/org/confeti/controllers/core/CompanyController.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                .onErrorResume(Exception.class, err -> Mono.just(ResponseEntity.badRequest().body(new ErrorResponse(err.getMessage()))));
+    }
+
+    @GetMapping(path = "/stat")
+    @ResponseBody
+    public Mono<ResponseEntity<?>> handleStatRequest() {
+        return reportStatsService.countCompanyStats()
+                .groupBy(ReportStatsByCompany::getCompanyName)
+                .flatMap(group ->
+                        group
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/core/CompanyController.java]: I think it's better `group.collectMap` and not
```
group
        .collectMap
```

File: api/src/main/java/org/confeti/service/dto/Company.java [hunk 3]
```diff
@@ -18,10 +20,12 @@
 @AllArgsConstructor(access = AccessLevel.PRIVATE)
 public class Company implements Serializable {
 
+    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
```
Comment (lapter57) [inline api/src/main/java/org/confeti/service/dto/Company.java]: Why?

File: api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                .expectStatus().isOk()
+                .expectBody(CompanyStatResponse.class).isEqualTo(expectedResponse);
+    }
+
+    @Test
+    public void testGetCompanyStatRequestWithoutYearForOneYear() throws JsonProcessingException {
+        final String companyName = "company-test";
+        final int year = 1972;
+        final long amount = 5L;
+        final ObjectMapper mapper = new ObjectMapper();
```
Comment (lapter57) [inline api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java]: You don't use it

File: api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                .build()
+                .get()
+                .uri(String.format("/api/rest/company/%s/stat?year=%d", companyName, year))
+                .exchange()
+                .expectStatus().isOk()
+                .expectBody(CompanyStatResponse.class).isEqualTo(expectedResponse);
+    }
+
+    @Test
+    public void testGetCompanyStatRequestWithoutYearForOneYear() throws JsonProcessingException {
```
Comment (lapter57) [inline api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java]: You can remove throw statement

File: api/src/main/java/org/confeti/service/dto/Company.java [hunk 6]
```diff
@@ -18,10 +20,12 @@
...
     private static final long serialVersionUID = 1L;
 
     private String name;
 
+    @JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (lapter57) [inline api/src/main/java/org/confeti/service/dto/Company.java:26]: I think it is not necessary

File: api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        when(companyService.findAll()).thenReturn(Flux.fromIterable(companies));
+
+        final EntityExchangeResult<List> response = WebTestClient
+                .bindToController(companyController)
+                .build()
+                .get()
+                .uri("/api/rest/company")
+                .exchange()
+                .expectStatus().isOk()
+                .expectBody(List.class).returnResult();
```
Comment (lapter57) [inline api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java]: You can just write like this and not use `ObjectMapper`:
```
        WebTestClient
                .bindToController(companyController)
                .build()
                .get()
                .uri("/api/rest/company")
                .exchange()
                .expectStatus().isOk()
                .expectBody(Company[].class).isEqualTo(companies.toArray(new Company[0]));
```

File: api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        );
+
+        final EntityExchangeResult<List> response = WebTestClient
+                .bindToController(companyController)
+                .build()
+                .get()
+                .uri("/api/rest/company/stat")
+                .exchange()
+                .expectStatus().isOk()
+                .expectBody(List.class).returnResult();
```
Comment (lapter57) [inline api/src/test/java/org/confeti/controllers/core/CompanyControllerTest.java]: You can just write like this and not use `ObjectMapper`:
```
        WebTestClient
                .bindToController(companyController)
                .build()
                .get()
                .uri("/api/rest/company/stat")
                .exchange()
                .expectStatus().isOk()
                .expectBody(CompanyStatResponse[].class).isEqualTo(statResponses.toArray(new CompanyStatResponse[0]));
```

File: api/src/main/java/org/confeti/controllers/core/CompanyController.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                .<ResponseEntity<?>>map(ResponseEntity::ok)
+                .onErrorResume(Exception.class, err -> Mono.just(ResponseEntity.badRequest().body(new ErrorResponse(err.getMessage()))));
+
+    }
+
+    @GetMapping
+    @ResponseBody
+    public Mono<ResponseEntity<?>> handleCompanyRequest() {
+        return companyService.findAll()
+                .collectList()
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/core/CompanyController.java:81]: I think it's better to return an object with field `List<Company> companies`

File: api/src/main/java/org/confeti/service/dto/Company.java (hunk 6 above)
Comment (marsofandrew) [reply api/src/main/java/org/confeti/service/dto/Company.java:26]: But looks good for users

File: api/src/main/java/org/confeti/controllers/core/CompanyController.java (hunk 9 above)
Comment (marsofandrew) [reply api/src/main/java/org/confeti/controllers/core/CompanyController.java:81]: No, usually in this case array should be returned

**Review by lapter57 — APPROVED** [review]

**Pull request merged**